components/plurals/src/rules/runtime/ast.rs

```diff
@@ -0,0 +1,634 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+#![allow(missing_docs)]
+
+use crate::rules::reference;
+use core::{convert::TryInto, fmt, mem, str::FromStr};
+use icu_provider::yoke::{self, *};
+use zerovec::{
+    ule::{custom::EncodeAsVarULE, AsULE, PlainOldULE, VarULE, ULE},
+    {VarZeroVec, ZeroVec},
+};
+
+#[derive(Debug, Yokeable, ZeroCopyFrom)]
+pub struct Rule<'data>(pub VarZeroVec<'data, RelationULE>);
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+#[repr(u8)]
+pub enum AndOr {
+    Or,
+    And,
+}
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug)]
+#[repr(u8)]
+pub enum Polarity {
+    Negative,
+    Positive,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+#[repr(u8)]
+pub enum Operand {
+    N,
+    I,
+    V,
+    W,
+    F,
+    T,
+    C,
+    E,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum RangeOrValue {
+    Range(u32, u32),
+    Value(u32),
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Relation<'data> {
+    pub(crate) and_or: AndOr,
+    pub(crate) polarity: Polarity,
+    pub(crate) operand: Operand,
+    pub(crate) modulo: u32,
+    pub(crate) range_list: ZeroVec<'data, RangeOrValue>,
+}
+
+/////
+
+impl Operand {
+    fn idx(self) -> u8 {
+        self as u8
+    }
+
+    fn from_idx(idx: u8) -> Self {
+        debug_assert!(idx < 8);
+        // safe due to the repr(u8)
+        unsafe { mem::transmute(idx) }
```

Comment:
Suggestion (optional): using `num_enum` would remove the `unsafe` from our code.  Also I'm not a big fan of having the `8` statically in our function with no way to tie it to the enum if we add new entries to the enum in the future.